File: C++/Algorithms/Sieve Algorithms/Sieve_of_Eratosthenes.cpp
/*
    Sieve of Eratosthenes:
    Given a number n, print all prime numbers less than n
*/

#include <iostream>
using namespace std;

void sieve_of_eratosthenes(int n)
{
    /*
    This will print all 
    prime number less than n
    */

    // Initially all are prime
    bool *prime = new bool[n];
    for (int i = 0; i < n; i++)
    {
        prime[i] = true;
    }

    // Now we start marking multiples of every number non-prime
    for (int i = 2; i * i <= n; i++)
    {
        // We start from first multiple and go till n
        for (int j = 2 * i; j < n; j += i)
        {
            prime[j] = false;
        }
    }
    for (int i = 2; i < n; i++)
    {
        if (prime[i])
        {
            cout << i << " ";
        }
    }
}

int main()
{
    sieve_of_eratosthenes(50);
}

File: Python/Algorithms/Sieve Algorithms/sieveOfEratosthenes.py
# Sieve of Eratosthenes
# It is an efficient algorithm used to find all the prime numbers smaller than
# or equal to a number(n)
# We will create a list and add all the elements and then remove those elements
#  which are not prime


def sieveOfEratosthenes(n):
    # Creating an empty list
    primeList = []

    for i in range(2, n + 1):
        # Adding all the elements
        primeList.append(i)

    i = 2
    while i * i <= n:
        if i in primeList:
            # Here we will remove all the elements that are multiples of i less
            # than n + 1 and greater than i
            for j in range(i * 2, n + 1, i):
                if j in primeList:
                    primeList.remove(j)

        i = i + 1

    print(primeList)


n = int(input("Enter the number upto which \
        you want to print the prime numbers :"))
sieveOfEratosthenes(n)
